bindings/bind_py_structure_manager.cc

```diff
@@ -25,489 +25,646 @@
-decltype(auto) add_iterator(
-    py::module & m,
-    PyClusterRef<StructureManagerImplementation, Order - 1> & py_cluster) {
-  // Order-1 for PyClusterRef because it is the
-  // cluster from the previous iteration
-  using Child = StructureManagerImplementation;
-  using Parent = typename Child::Parent;
-
-  // bind the iteration over clusterRef<Order-1>
-  using ClusterRef = typename Parent::template ClusterRef<Order - 1>;
-  py_cluster.def(
-      "__iter__",
-      [](ClusterRef & v) { return py::make_iterator(v.begin(), v.end()); },
-      py::keep_alive<0, 1>()); /* Keep vector alive while iterator is used */
-  auto py_cluster_new = add_cluster<Order, Child>(m);
-  return py_cluster_new;
-}
-
-//! bind iterator and ClusterRef for Order == 1
-template <typename StructureManagerImplementation, size_t Order>
-decltype(auto)
-add_iterator(py::module & m,
-             PyManager<StructureManagerImplementation> & manager) {
-  using Child = StructureManagerImplementation;
-  using Parent = typename Child::Parent;
-
-  // bind the iteration over the manager
-  manager.def(
-      "__iter__",
-      [](Parent & v) { return py::make_iterator(v.begin(), v.end()); },
-      py::keep_alive<0, 1>()); /* Keep vector alive while iterator is used */
-  auto py_cluster = add_cluster<1, Child>(m);
-  return py_cluster;
-}
-
-/**
- * Bind the clusterRef allowing to iterate over the manager, atom, neigh...
- * Use signature overloading to dispatch to the proper function.
- * Use iteration by recursion to iterate from Order to MaxOrder-1 staticaly
- */
-template <typename StructureManagerImplementation, size_t Order,
-          size_t MaxOrder>
-struct add_iterators {
-  //! starts recursion
-  static void static_for(py::module & m,
-                         PyManager<StructureManagerImplementation> & manager) {
-    auto py_cluster =
-        add_iterator<StructureManagerImplementation, Order>(m, manager);
-    add_iterators<StructureManagerImplementation, Order + 1,
-                  MaxOrder>::static_for(m, py_cluster);
+  //! Bind many cluster refs of a given order using template recursion
+  template <size_t Order, size_t Layer, size_t LayerEnd>
+  struct add_cluster_refs {
+    //! starts recursion
+    static void static_for(py::module & m) {
+      add_cluster_ref<Order, Layer>(m);
+      add_cluster_refs<Order, Layer + 1, LayerEnd>::static_for(m);
+    }
+  };
+
+  //! Stop the recursion
+  template <size_t Order, size_t LayerEnd>
+  struct add_cluster_refs<Order, LayerEnd, LayerEnd> {
+    static void static_for(py::module &) {}
+  };
+
+  //! Bind clusters of different orders
+  //! (up to quadruplet)
+  template <size_t Order, typename SMI>
+  decltype(auto) add_cluster(py::module & m) {
+    using ClusterRef = ClusterRef_t<SMI, Order>;
+    using ClusterRefKey = ClusterRefkey_t<SMI, Order>;
+
+    std::string cluster_name = internal::GetBindingTypeName<SMI>();
+
+    static_assert(Order < 5, "Can't bind more than quadruplet as it is.");
```

Comment:
Wording -- I'd say `"Not currently configured to bind clusters higher than quadruplets."`, which implies that the bound cluster order can be (relatively) easily increased later.